File: array/197.cpp
#include <bits/stdc++.h>

using namespace std;

void moveZeroes(vector<int>& nums) {
    int L = 0, R = 1;

    while(R < nums.size()) {
        if(nums[L] != 0) {
            ++L;
            R = L + 1;
            continue;
        }
        if(nums[R] != 0) {
            swap(nums[L], nums[R]);
            ++L;
            ++R;
        } else {
            ++R;
        }
    }
}

int main() {
    vector<int> nums = {0,1,0,3,12};
    moveZeroes(nums);

    for(const int &e: nums) printf("%d ", e);


    return 0;
}

File: array/197.py
from typing import List


def moveZeroes(nums: List[int]) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """

    L, R = 0, 1

    while R < len(nums):
        if nums[L] != 0:
            L += 1
            R = L + 1
            continue
        if nums[R] != 0:
            nums[L], nums[R] = nums[R], nums[L]
            R += 1
            L += 1
        else:
            R += 1


def moveZeroes2(nums: List[int]):
    nums.sort(key=lambda x: x == 0)


nums = [0, 1, 0, 3, 12]
moveZeroes2(nums)

print(nums)
